gateway/crates/gateway-binary/src/telemetry.rs

```diff
@@ -15,25 +15,56 @@ pub(crate) struct OpenTelemetryProviders {
 }
 
 pub(crate) fn init(args: &impl Args, config: TelemetryConfig) -> anyhow::Result<OpenTelemetryProviders> {
+    use grafbase_telemetry::otel::opentelemetry::propagation::TextMapPropagator;
+    use grafbase_telemetry::otel::opentelemetry_sdk::trace::RandomIdGenerator;
+    use opentelemetry_aws::trace::XrayPropagator;
     use tracing_subscriber::layer::SubscriberExt;
     use tracing_subscriber::util::SubscriberInitExt;
 
     let filter = args.log_level().map(|l| l.as_filter_str()).unwrap_or("info");
     let env_filter = EnvFilter::new(filter);
-    let id_generator = {
-        cfg_if::cfg_if! {
-            if #[cfg(feature = "lambda")] {
-                use opentelemetry_aws::trace::{XrayIdGenerator, XrayPropagator};
-                grafbase_telemetry::otel::opentelemetry::global::set_text_map_propagator(XrayPropagator::default());
-
-                XrayIdGenerator::default()
-            } else {
-                use grafbase_telemetry::otel::opentelemetry_sdk::trace::RandomIdGenerator;
-
-                RandomIdGenerator::default()
-            }
+
+    {
+        let mut propagators: Vec<Box<dyn TextMapPropagator + Send + Sync>> = Vec::new();
+
+        let mut xray_propagator_added = false;
+
+        let tracing_config = &config.tracing;
+        if tracing_config.propagation.trace_context {
+            propagators.push(Box::new(
+                grafbase_telemetry::otel::opentelemetry_sdk::propagation::TraceContextPropagator::new(),
+            ));
+        }
+
+        if tracing_config.propagation.baggage {
+            propagators.push(Box::new(
+                grafbase_telemetry::otel::opentelemetry_sdk::propagation::BaggagePropagator::new(),
+            ))
+        }
+
+        if tracing_config.propagation.aws_xray {
+            propagators.push(Box::new(XrayPropagator::default()));
+            xray_propagator_added = true;
         }
-    };
+
+        if cfg!(feature = "lambda") && !xray_propagator_added {
+            propagators.push(Box::new(XrayPropagator::default()))
```

Comment:
Should we get rid of this and use the config option only?

Reply:
That would work, a bit more configuration burden for the user but simpler for us, and less branching.

Reply:
exactly. nobody uses it on lambda yet, so we can break things.

Reply:
Ok doing that then.